### ISSUE TITLE
Support overload implementation consistency with generics

### ISSUE THREAD

**Issue by charliermarsh** [astral-sh/ty#3511]
In https://github.com/astral-sh/ruff/pull/24936, we added support for validating overload implementation consistency, as in:

```python
from typing import overload

@overload
def g(x: int) -> int: ...
@overload
def g(x: str) -> str: ...

def g(x: int | str) -> int:
    return 1
```

However, that implementation is limited to signatures and implementations that omit generics. We need to extend support to generics, as in:

```python
from typing import TypeVar, overload

T = TypeVar("T")

@overload
def g(x: list[T]) -> list[T]: ...
@overload
def g(x: None) -> None: ...

def g(x: list[T] | None) -> list[int] | None: ...
```

**Comment by thomasaarholt** [comment]
The example at the bottom of this issue passes ty check, and I think it can be closed!
[Playground](https://play.ty.dev/18623431-4d0f-4b02-9bc8-44c75cfb8ada)

I noticed this while digging for similar issues as one I am reporting.

**Comment by AlexWaygood** [comment]
> The example at the bottom of this issue passes ty check

indeed -- we emit no diagnostic currently, but we should do :-) Other type checkers [such as mypy](https://mypy-play.net/?gist=07e009cc975a6ccd49d72a53e542a815) emit a diagnostic here. This issue exists to track the false negative from ty. So no, this issue cannot be closed yet.

**Comment by thomasaarholt** [comment]
Whoops! I didn't realize that it was that way around! Sorry for the noise! 😅
